### ISSUE TITLE
jupyter2: disable specific file → exports if underlying software isn't available

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#4093]
i.e. use the capabilities configuration object to disable certain menu entries or buttons. in particular, if xelatex is missing, it's not possible to export a PDF via latex  in jupyter's file menu.

**Comment by williamstein** [comment]
We might want to show the option, but if it is selected and related isn't
available, we tell the user they need to install xelatex.

This is fundamentally different than the "totally hide the functionality"
approach you have been taking.  However it is common with Linux apps that
I've used in the past...  it also avoids confusion about *where is this
feature?* and is of course closer to what Jupiter official does (which is
never hides stuff).

**Comment by williamstein** [comment]
I think xelatex is only 9MB, so it should always be installed...

**Comment by haraldschilly** [comment]
well, if the custom software environment doesn't define it to be installed, we should always install it regardless of that?

despite that, xelatex alone doesn't make any ipynb to pdf via latex export work. it needs a lot of other styles, fonts and latex packages.

**Comment by williamstein** [comment]
>  **a** custom software environment

I completely forgot about that, where there are a million possibilities for how things can be broken, and you can't install anything.  Of course.   Is that what the user in zendesk is asking about?

**Comment by haraldschilly** [comment]
yes, I made a PR but it is rejected: https://github.com/AllenDowney/ModSimPy/pull/47

this is one of the cases where high level requests can either be solved by "fixing it properly" in the underlying layer or adding something as some sort of workaround. the drawback with the workaround solution is to introduce overhead (extra work) down the road or even worse, sudden inconsistencies and breaks in the future. that's why I tried to push this upstream.

**Comment by williamstein** [comment]
Your screenshot caused the PR to be "unrejected": " Ok, let's give this a
try.".  Yeah!

**Comment by haraldschilly** [comment]
yes, very good. In an ideal world, repo2docker (used by us and mybinder) would always make sure all jupyter notebook capabilities are available. it doesn't, and I understand why: always dockerizing a latex environment is heavy. In any case, that PR helps to fill in the gap.

 :rainbow: :stars:  :rainbow:  dreaming of a better world :rainbow: :stars:  :rainbow:

**Comment by williamstein** [comment]
With compute servers, there are no lots of little environments where (1) things like nbconvert and xelatex are definitely not installed, and (2) users can install them fairly easily (as they are root).   

I think the right solution is to just leave the option there, and when you click on it, clearly surface the error that caused things to not work.  The user can then decide to install, ask us, or try something else.  This is exactly what happens right now:

<img width="894" alt="image" src="https://github.com/user-attachments/assets/d9fe3ccd-e5a8-43ce-9eed-ca30e272a334">

.... 

except evidently with compute servers, right now *both* things happen at once:

- you get an error that stuff isn't installed, and 
- you ALSO get the nicely printed out result.

I think probably the print is happening on both the the compute server and  in the home base at the same time, and it works on one but not the other.  

In any case, I'm marking this issue as "unclear", since I disagree with entirely removing menu items when functionality isn't available, but could easily be made available by the user doing something.   With compute servers, you can even auto install a dependency, or at least offer to do so -- that's how the JupyterLab and VS Code launchers for compute servers already work.  I could have implemented compute servers by saying: "let's detect if jupyterlab is installed; if not, do not even show a juypterlab launcher button".  That's your approach.  Instead I implemented them as follows: "show a jupyterlab button; when you click it, install jupyterlab if necessary".
